console-subscriber/src/visitors.rs

```diff
@@ -0,0 +1,359 @@
+use super::{AttributeUpdate, AttributeUpdateOp, AttributeUpdateValue, Readiness, WakeOp};
+use console_api as proto;
+use proto::resources::resource;
+use std::collections::HashMap;
+use tracing_core::{
+    field::{self, Visit},
+    span,
+};
+
+#[derive(Default)]
+pub(crate) struct ResourceVisitor {
+    concrete_type: Option<String>,
+    kind: Option<resource::Kind>,
+}
+
+pub(crate) struct FieldVisitor {
+    fields: Vec<proto::Field>,
+    meta_id: proto::MetaId,
+}
+
+#[derive(Default)]
+pub(crate) struct AsyncOpVisitor {
+    source: Option<String>,
+}
+
+#[derive(Default)]
+pub(crate) struct WakerVisitor {
+    id: Option<span::Id>,
+    op: Option<WakeOp>,
+}
+
+#[derive(Default)]
+pub(crate) struct ResourceOpVisitor {
+    op_name: Option<String>,
+    op_type: Option<String>,
+    readiness: Option<Readiness>,
+    state_text_attrs: HashMap<String, String>,
+    state_numeric_attrs: HashMap<String, NumericStateAttr>,
+}
+
+#[derive(Debug)]
+pub(crate) enum ResourceOpData {
+    Poll {
+        op_name: String,
+        readiness: Readiness,
+    },
+    StateUpdate {
+        op_name: String,
+        attrs: Vec<AttributeUpdate>,
+    },
+}
+
+#[derive(Default, Debug)]
+struct NumericStateAttr {
+    val: Option<u64>,
+    op: Option<String>,
+    unit: Option<String>,
+}
+
+impl ResourceVisitor {
+    const RES_CONCRETE_TYPE_FIELD_NAME: &'static str = "concrete_type";
+    const RES_KIND_FIELD_NAME: &'static str = "kind";
+    const RES_KIND_TIMER: &'static str = "timer";
+
+    pub(crate) fn result(self) -> Option<(String, resource::Kind)> {
+        self.concrete_type.zip(self.kind)
+    }
+}
+
+impl Visit for ResourceVisitor {
+    fn record_debug(&mut self, _: &field::Field, _: &dyn std::fmt::Debug) {}
+
+    fn record_str(&mut self, field: &tracing_core::Field, value: &str) {
+        if field.name() == Self::RES_CONCRETE_TYPE_FIELD_NAME {
+            self.concrete_type = Some(value.to_string());
+        } else if field.name() == Self::RES_KIND_FIELD_NAME {
+            let kind = Some(match value {
+                Self::RES_KIND_TIMER => {
+                    resource::kind::Kind::Known(resource::kind::Known::Timer as i32)
+                }
+                other => resource::kind::Kind::Other(other.to_string()),
+            });
+            self.kind = Some(resource::Kind { kind });
+        }
```

Comment:
nit/tioli: maybe use 
```suggestion
        match field.name() {
            Self::RES_CONCRETE_TYPE_FIELD_NAME => self.concrete_type = Some(value.to_string()),
            Self::RES_KIND_FIELD_NAME => {
                let kind = Some(match value {
                    Self::RES_KIND_TIMER => {
                        resource::kind::Kind::Known(resource::kind::Known::Timer as i32)
                    }
                    other => resource::kind::Kind::Other(other.to_string()),
                });
                self.kind = Some(resource::Kind { kind });
            },
            _ => {},
        }
```